CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+```
+
+For `builtinLogicalXor`, we instead have (again, irrespective of semantics):
+
+```haskell
+builtinLogicalXor s x (builtinLogicalComplement x) = x
+```
+
+#### Bit reading and modification
+
+Throughout, we assume any index arguments to be 'in-bounds'; that is, all the
+index arguments used in the statements of any law are such that the operation
+they are applied to wouldn't produce an error.
+
+The first law of `builtinSetBits` is similar to the [set-twice law of
+lenses][lens-laws]:
+
+```haskell
+builtinSetBits bs [(i, b1), (i, b2)] = builtinSetBits bs [(i, b2)]
+```
+
+Together with `builtinReadBit`, we obtain the remaining two analogues to the lens
+laws:
+
+```haskell
+-- writing to an index, then reading from that index, gets you what you wrote
+builtinReadBit (builtinSetBits bs [(i, b)]) i = b
+
+-- if you read from an index, then write that value to that same index, nothing
+-- happens
+builtinSetBits bs [(i, builtinReadBit bs i)] = bs
+```
+
+Furthermore, given a fixed data argument, `builtinSetBits` acts as a [monoid
+homomorphism][monoid-homomorphism] from functions to lists under concatenation:
+
+```haskell
+-- identity function corresponds to empty list
+builtinSetBits bs [] = bs
+
+-- composition is concatenation of change list arguments
+builtinSetBits (builtinSetBits bs is) js = builtinSetBits bs (is <> js)
+```
+
+#### `builtinReplicate`
+
+Given a fixed byte argument, `builtinReplicate` acts as a [monoid
+homomorphism][monoid-homomorphism] from natural numbers under addition to
+`BuiltinByteString`s under concatenation: 
+
+```haskell
+builtinReplicate 0 w = ""
+
+builtinReplicate (n + m) w = builtinReplicate n w <> builtinReplicate m w
+```
+
+Additionally, for any 'in-bounds' index (that is, any index for which
+`builtinIndexByteString` won't error) `i`, we have
+
+```haskell
+builtinIndexByteString (builtinReplicate n w) i = w
+```
+
+Lastly, we have
+
+```haskell
+builtinSizeOfByteString (builtinReplicate n w) = n
+```
+
+## Rationale: how does this CIP achieve its goals?
+
+The operations, and semantics, described in this CIP provide a set of
+well-defined bitwise logical operations, as well as bitwise access and
+modification, to allow cases similar to Case 1 to be performed efficiently and
+conveniently. Furthermore, the semantics we describe would be reasonably
+familiar to users of other programming languages (including Haskell) which have
+provisions for bitwise logical operations of this kind, as well as some way of
+extending these operations to operate on packed byte vectors. At the same time,
+there are several choices we have made that are somewhat unusual, or could
+potentially have been implemented differently based on existing work: most
+notably, our choice of bit indexing scheme, the padding-versus-truncation
+semantics, and the multiplicitous definition of bit modification. Among existing
+work, a particularly important example is [CIP-58][cip-58], which makes
```

Comment:
One thing I'm missing is a discussion of the things that are not included from CIP-58. It's unclear to me whether the intention is simply for this to be another tranche of that work, or whether we've decided that some of the CIP-58 operations are less useful.

Reply:
This is meant to be a continuation of CIP-121, which was meant to break 58 into manageable chunks, as well as to address some of the shortcomings of the original work. I should also make this more clear.